### DISCUSSION TITLE
full ir support + power sensor not configurable

### DISCUSSION THREAD

**Discussion by HMvdZijden** [arendst/Tasmota#14901] [Q&A] [answered]
### PROBLEM DESCRIPTION
_A clear and concise description of what the problem is._
I try to build tasmota with full IR support by activating line #define USE_IR_REMOTE_FULL                     // Support all IR protocols from IRremoteESP8266
After that the compile ends in an error.
Linking .pio/build/tasmota/firmware.elf
/home/gitpod/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld: .pio/build/tasmota/src/tasmota.ino.cpp.o:(.text._Z14IrRemoteSendGCPPcjj+0x0): undefined reference to `IRsend::sendGC(unsigned short*, unsigned short)'
/home/gitpod/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld: .pio/build/tasmota/src/tasmota.ino.cpp.o: in function `IrRemoteSendGC(char**, unsigned int, unsigned int)':
tasmota.ino.cpp:(.text._Z14IrRemoteSendGCPPcjj+0x7d): undefined reference to `IRsend::sendGC(unsigned short*, unsigned short)'
/home/gitpod/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld: .pio/build/tasmota/src/tasmota.ino.cpp.o:(.text._Z23IrRemoteSendRawStandardPPctjj+0x0): undefined reference to `IRsend::sendRaw(unsigned short const*, unsigned short, unsigned short)'
/home/gitpod/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld: .pio/build/tasmota/src/tasmota.ino.cpp.o: in function `IrRemoteSendRawStandard(char**, unsigned short, unsigned int, unsigned int)':
tasmota.ino.cpp:(.text._Z23IrRemoteSendRawStandardPPctjj+0x7b): undefined reference to `IRsend::sendRaw(unsigned short const*, unsigned short, unsigned short)'
/home/gitpod/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld: .pio/build/tasmota/src/tasmota.ino.cpp.o: in function `IrRemoteSendRawFormatted(char**, unsigned int, unsigned int)':
tasmota.ino.cpp:(.text._Z24IrRemoteSendRawFormattedPPcjj+0x10e): undefined reference to `IRsend::sendRaw(unsigned short const*, unsigned short, unsigned short)'
/home/gitpod/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld: tasmota.ino.cpp:(.text._Z24IrRemoteSendRawFormattedPPcjj+0x1eb): undefined reference to `IRsend::sendRaw(unsigned short const*, unsigned short, unsigned short)'
collect2: error: ld returned 1 exit status
*** [.pio/build/tasmota/firmware.elf] Error 1

Building tasmota-ir builds one with full support for ir but does not support power sensors and I need that combination.

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): _____NA
- [x] Tasmota binary firmware version number used: _____latest
  - [ ] Pre-compiled
  - [x] Self-compiled
- [ ] Flashing tools used: _____
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:


```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Steps to reproduce the behavior:_
Uncomment  line 836 in my_user_config_h

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._
A build output

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by barbudor** [comment]
Delete the `.pio`, the `.cache` and the `$HOME/.platformio` folders and let PlatformIO redownload fresh and clean environment

**Comment by Jason2866** [comment]
It does NOT work this way. You have to define your own env and place it in `platformio_tasmota_cenv.ini`
```
[env:tasmota-HMvdZijden]
build_flags = ${env.build_flags} -DUSE_IR_REMOTE_FULL
```
with `pio run -e tasmota-HMvdZijden` you get your Tasmota version with Full IR support

**Comment by barbudor** [comment]
@Jason2866 So the instructions are incomplete as they do not refer to a custom env

https://github.com/arendst/Tasmota/blob/202f83b4ff14d8fc7eec79662b8fa82d8fa1a9a6/tasmota/my_user_config.h#L835-L842

We probably should write something here in my_user_config
Do you want me to do it ?

**Comment by Jason2866** [comment]
@barbudor  yes, please. I am on Mobile now.

**Comment by HMvdZijden** [comment]
I am really sorry, i am such a noob, I really don’t know what to do and why.
Can someone inform me what to do exactly.

**Comment by Jason2866** [comment]
@HMvdZijden  No reason for excuses! We all started at point zero.
So what do you want to have included in your firmware?

**Comment by HMvdZijden** [comment]
Only power management and irhavac for mitsubishi. I have an ac unit of which I want to see usage and control his functions by ir.
So I was editing my_user_config eliminating all not necessary sensors and features because I understood that full ir not fits in standard Tasmota

**Comment by Jason2866** [comment]
Nearly there. Recommended is NOT to edit `my_userconfig.h` for user edits.
Copy file `user_config_override_sample.h` to `user_config_override.h` and do the (un)defines there.
Using the env
```
[env:tasmota-HMvdZijden]
build_flags = ${env.build_flags} -DUSE_IR_REMOTE_FULL
```
will include this changes

**Comment by HMvdZijden** [comment, marked as answer]
I started a new workspace, so everything is fresh.
Then I added 
[env:tasmota-HMvdZijden]
build_flags = ${env.build_flags} -DUSE_IR_REMOTE_FULL
to file platformio_tasmota_cenv.ini and did the suggested run.
No problem at all, without undef anything.
Still don't understand why but it workes. 
Thank you Jason and Barbudor.

**Comment by HMvdZijden** [comment]
Got another problem now. The build I uploaded is that big that even een minimal build cannot be uploaded anymore because there is to little space left. I cannot flash over serial with this device, so I cannot update it anymore.

**Reply by ascillato** [reply]
Please, disable unused features to reduce the file size.

**Comment by barbudor** [comment]
Did you used the GZ file?
For ota/upload always use the GZ

**Comment by HMvdZijden** [comment]
The tasmota-minimal.bin.gz is too big😢

**Reply by sfromis** [reply]
There is a smaller -minimal here, without web server, tasmota-minicustom. Be really, really careful, and know what you're doing.... Options are low if you screw it up.
https://github.com/Jason2866/Tasmota-specials

Another possibility (not tested this) could be to replace (not upgrade) firmware via https://github.com/Jason2866/ESP8266MinimalHTTPUpdater

Anyway, wired flashing always works.....

**Reply by HMvdZijden** [reply]
Can you tell me the maximum size of a build to always be able to ota upgrade with minimal?

**Reply by ascillato** [reply]
670Kb approx. using minimal as intermediate in a 1MB flash device.

**Reply by HMvdZijden** [reply]
I used the small web update. Fantastic, no problems anymore. My current bin is small enough to be able to keep updating. Maybe it would be nice if you mention on this software that it cannot be used with gz files.
Thanks anyway for your engagement in this project.

**Reply by ascillato** [reply]
Sorry, gz files are indeed supported for updating Tasmota by OTA.

**Comment by HMvdZijden** [comment]
I reflashed with minimalhttpUpdater. Succeeded,
Maybe you best write in the documentation that it cannot use .gz files.

I am happy and keep this tool aside. Thank You very much.

**Comment by bsoerni** [comment]
Hello to all, first I am new here and no crack with tasmota and esp programming. I have flashed some magic home wlan led controller (yes i had five or six times success). I flashed the wlan only AND the ir variants.
The Wlan functionality is full working light on/off color change. But IR does not what i expected.
In the console windows i can see the messsages but the thing does nothing. I tried some "rules" but nothing happen. I hope it is possible that the wlan controller does on/off and color change and dimming without mqtt and openhab (thats the system i use).

Many thanks.

P.S.: Hope it's the right place for this question

**Reply by barbudor** [reply]
Flashing Tasmota will not "magically" link the IR to the light control
You have to build that link using rules.
What do you see in the console when pressing buttons on the remote ?

**Reply by sfromis** [reply]
Here's an example of linking IR input with actions on the same device.
https://templates.blakadder.com/MagicHome_ESP-IR-B-v2_3

**Reply by bsoerni** [reply]
Wow your're fast! Thanks!
First i will answer what i see in console:

15:54:02.034 RSL: RESULT = {"Time":"2022-04-20T15:54:02","IrReceived":{"Protocol":"NEC","Bits":32,"Data":"0x00FFB04F"}}
15:54:02.182 RSL: RESULT = {"Time":"2022-04-20T15:54:02","IrReceived":{"Protocol":"NEC","Bits":0,"Data":"0xFFFFFFFFFFFFFFFF"}}
15:54:04.581 RSL: RESULT = {"Time":"2022-04-20T15:54:04","IrReceived":{"Protocol":"NEC","Bits":32,"Data":"0x00FFF807"}}
15:54:04.733 RSL: RESULT = {"Time":"2022-04-20T15:54:04","IrReceived":{"Protocol":"NEC","Bits":0,"Data":"0xFFFFFFFFFFFFFFFF"}}

The first two lines are button press on and second two lines the same for off button


AND i tried your solution @sfromis thanx this worked! I could not say - at the moment - why the other rules i tried did not work. But this set of rules does nearly 100% of what i expect. I think the rest will be a little trial and error.

can yout tell me what i can do to replicate this settings to my other Led Controller? And/or more important how can i make these rules persistent (upgrading or new installation directly with these settings)

**Reply by sfromis** [reply]
Rule sets are persistent unless you explicitly erase. Using the command `Rule0`, you can get current rule sets, and copy/paste to another device. Anyway, I'd suggest creating+saving the rule sets outside of Tasmota, and then paste the commands into a device when you need it.
